crates/datasources/src/sqlserver/mod.rs

```diff
@@ -0,0 +1,559 @@
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<DatafusionResult<RecordBatch>>> {
+        loop {
+            match &mut self.stream_state {
+                RowStreamState::Opening { opening_fut } => {
+                    match ready!(opening_fut.poll_unpin(cx)) {
+                        Ok(stream) => {
+                            // We have the stream, advance state.
+                            let stream = Box::pin(stream.chunks(self.chunk_size));
+                            self.stream_state = RowStreamState::Scan { stream };
+                            continue;
+                        }
+                        Err(e) => {
+                            self.stream_state = RowStreamState::Done;
+                            return Poll::Ready(Some(Err(DataFusionError::External(Box::new(e)))));
+                        }
+                    };
+                }
+
+                RowStreamState::Scan { stream } => match ready!(stream.poll_next_unpin(cx)) {
+                    Some(chunk) => match rows_to_record_batch(chunk, self.arrow_schema.clone()) {
+                        Ok(batch) => return Poll::Ready(Some(Ok(batch))),
+                        Err(e) => {
+                            self.stream_state = RowStreamState::Done;
+                            return Poll::Ready(Some(Err(DataFusionError::External(Box::new(e)))));
+                        }
+                    },
+                    None => self.stream_state = RowStreamState::Done,
+                },
+
+                RowStreamState::Done => return Poll::Ready(None),
+            }
+        }
+    }
+}
+
+impl Stream for RowStream {
+    type Item = DatafusionResult<RecordBatch>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.poll_next_inner(cx)
+    }
+}
+
+impl RecordBatchStream for RowStream {
+    fn schema(&self) -> ArrowSchemaRef {
+        self.arrow_schema.clone()
+    }
+}
+
+/// Convert a chunk of rows into a record batch.
+fn rows_to_record_batch(
+    rows: Vec<Result<tiberius::Row>>,
+    schema: ArrowSchemaRef,
+) -> Result<RecordBatch> {
+    use datafusion::arrow::array::{
+        Array, BinaryBuilder, BooleanBuilder, Float32Builder, Float64Builder, Int16Builder,
+        Int32Builder, Int64Builder, StringBuilder, TimestampNanosecondBuilder,
+    };
+
+    let rows = rows.into_iter().collect::<Result<Vec<_>>>()?;
+
+    /// Macro for generating the match arms when converting rows to a record batch.
+    macro_rules! make_column {
+        ($builder:ty, $rows:expr, $col_idx:expr) => {{
+            let mut arr = <$builder>::with_capacity($rows.len());
+            for row in $rows.iter() {
+                arr.append_option(row.try_get($col_idx)?);
+            }
+            Arc::new(arr.finish())
+        }};
+    }
+
+    let mut columns = Vec::with_capacity(schema.fields.len());
+    for (col_idx, field) in schema.fields.iter().enumerate() {
```

Comment:
seems like there are more types in the schema that we reference, but by the time we're here we've already done the more messy cast.

Reply:
Only think I'm missing (after double checking) is the null type.

It's important that we convert to the exact type that's in the schema that we generated above since everything was planned with that schema. If we hit something here where we can't do the right cast, then the schema isn't correct. We can't say a column is an `Int32` then return an array of `Int64` since there's expectations downstream about the type.